src/Umbraco.SampleSite/Controllers/LessonsController.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Runtime.Serialization;
+using System.Threading.Tasks;
+using System.Web.Http;
+using Newtonsoft.Json;
+using Umbraco.Core;
+using Umbraco.Core.Cache;
+using Umbraco.Core.Configuration;
+using Umbraco.Core.Logging;
+using Umbraco.Web.Editors;
+using Umbraco.Web.Mvc;
+using Umbraco.Web.WebApi.Filters;
+
+namespace Umbraco.SampleSite.Controllers
+{
+    [PluginController("Starterkit")]
+    public class LessonsController : UmbracoAuthorizedJsonController
+    {
+        /// <summary>
+        /// Fetches available lessons for a given section from our.umbaco.org
+        /// </summary>
+        /// <param name="path">Name of the documentation section to fetch from, ex: "getting-started", "Tutorials/Starter-kit/Lessons" </param>
+        /// <param name="baseUrl">The url to retrieve from, by default https://our.umbraco.org</param>
+        /// <returns></returns>
+        [ValidateAngularAntiForgeryToken]
+        public async Task<IEnumerable<Lesson>> GetLessons(string path, string baseUrl = "https://our.umbraco.org")
+        {
+            //We need the umbraco context to fetch the currrent user and version
+            var context = UmbracoContext;
+
+            //this only works in the context of a umbraco backoffice request
+            if (context == null)
+                throw new HttpResponseException(HttpStatusCode.InternalServerError);
+
+            //information for the request, so we could in the future filter by user, allowed sections, langugae and user-type
+            var user = Security.CurrentUser;
+            var userType = user.UserType.Alias;
+            var allowedSections = string.Join(",", user.AllowedSections);
+            var language = user.Language;
+            var version = UmbracoVersion.GetSemanticVersion().ToSemanticString();
+
+            //construct the url and cache key
+            var url = string.Format(baseUrl + "/Umbraco/Documentation/Lessons/GetDocsForPath?path={0}&userType={1}&allowedSections={2}&lang={3}&version={4}", path, userType, allowedSections, language, version);
+            var key = "umbraco-lessons-" + userType + language + allowedSections.Replace(",", "-") + path;
+
+            //try and find an already cached version of this request
+            var content = ApplicationContext.ApplicationCache.RuntimeCache.GetCacheItem<List<Lesson>>(key);
```

Comment:
You should use `RuntimeCache.GetOrAdd instead` as this will be much simpler

Reply:
I can't seem to find the method `GetOrAdd` on the `ApplicationContext.ApplicationCache.RuntimeCache`

In this in newer versions of Umbraco @Shazwazza ? If so can you point me to a usage ideally please :)

Reply:
Sorry it's called `GetCacheItem` which is a Get Or Add functionality. That is the standard way to use the runtime cache. It's an extension method in namespace Umbraco.Core.Cache